CMakeLists.txt

```diff
@@ -50,22 +50,54 @@ ADD_SUBDIRECTORY(wave_utils)
 # wave_containers: header-only library
 ADD_SUBDIRECTORY(wave_containers)
 
-# libwave.a
-ADD_CUSTOM_TARGET(
-    libwave ALL
-    COMMAND
-        ${CMAKE_AR} rc libwave.a
-            ${CMAKE_BINARY_DIR}/wave_controls/libwave_controls.a
-	    ${CMAKE_BINARY_DIR}/wave_geometry/libwave_geometry.a
-            ${CMAKE_BINARY_DIR}/wave_kinematics/libwave_kinematics.a
-            ${CMAKE_BINARY_DIR}/wave_matching/libwave_matching.a
-            ${CMAKE_BINARY_DIR}/wave_optimization/libwave_optimization.a
-            ${CMAKE_BINARY_DIR}/wave_utils/libwave_utils.a
-    DEPENDS
-        wave_controls
-	wave_geometry
-        wave_kinematics
-        wave_matching
-        wave_optimization
-        wave_utils
-)
+# libwave
+SET(LIBWAVE_MODULE_LIBS
+    ${CMAKE_BINARY_DIR}/wave_controls/libwave_controls.a
+    ${CMAKE_BINARY_DIR}/wave_geometry/libwave_geometry.a
+    ${CMAKE_BINARY_DIR}/wave_kinematics/libwave_kinematics.a
+    ${CMAKE_BINARY_DIR}/wave_matching/libwave_matching.a
+    ${CMAKE_BINARY_DIR}/wave_optimization/libwave_optimization.a
+    ${CMAKE_BINARY_DIR}/wave_utils/libwave_utils.a)
+
+SET(LIBWAVE_MODULES
+    wave_controls
+    wave_geometry
+    wave_kinematics
+    wave_matching
+    wave_optimization
+    wave_utils)
+
+IF(BUILD_SHARED_LIBS)
+    # SHARED LIBRARY
+    ADD_CUSTOM_TARGET(
+        libwave ALL
+        COMMAND
+            gcc -shared -o libwave.so
+              -Wl,--whole-archive
+                ${LIBWAVE_MODULE_LIBS}
+              -Wl,--no-whole-archive
+        DEPENDS
+            ${LIBWAVE_MODULES}
+    )
+    INSTALL(FILES ${CMAKE_BINARY_DIR}/libwave.so DESTINATION ${CMAKE_INSTALL_PREFIX}/lib)
+
+ELSE ()
+    # STATIC LIBRARY
+    ADD_CUSTOM_TARGET(
+        libwave ALL
+        COMMAND
+            ar -rcT libwave.a ${LIBWAVE_MODULE_LIBS}
+        DEPENDS
+            ${LIBWAVE_MODULES}
+    )
+    INSTALL(FILES ${CMAKE_BINARY_DIR}/libwave.a DESTINATION CMAKE_INSTALL_PREFIX}/lib)
+
+ENDIF()
+
+INSTALL(DIRECTORY wave_containers/include/ DESTINATION ${CMAKE_INSTALL_PREFIX}/include)
```

Comment:
To clarify, use `INSTALL(DIRECTORY wave_containers/include/ DESTINATION include)`

Relative paths are relative to `CMAKE_INSTALL_PREFIX`.